cluster-autoscaler/processors/nodeinfos/podtemplates/node_info_with_podtemplates_processors.go

```diff
@@ -0,0 +1,172 @@
+
+	ctx        context.Context
+	cancelFunc func()
+}
+
+// Process returns unchanged nodeInfos.
+func (p *nodeInfoWithPodTemplateProcessor) Process(ctx *ca_context.AutoscalingContext, nodeInfosForNodeGroups map[string]*schedulerframework.NodeInfo) (map[string]*schedulerframework.NodeInfo, error) {
+	// here we can use empty snapshot, since the NodeInfos that will be updated
+	// are from CloudProvider NodeTemplates.
+	clusterSnapshot := simulator.NewBasicClusterSnapshot()
+
+	// retrieve only once the podTemplates list.
+	// This list will be used for each NodeGroup.
+	podTemplates, err := p.podTemplateLister.List(labels.Everything())
+	if err != nil {
+		return nil, errors.ToAutoscalerError(errors.InternalError, err)
+	}
+
+	result := make(map[string]*schedulerframework.NodeInfo, len(nodeInfosForNodeGroups))
+	var errs []error
+	for id, nodeInfo := range nodeInfosForNodeGroups {
+		var newNodeInfo *schedulerframework.NodeInfo
+
+		// only runs getNodeInfoWithPodTemplates() for NodeInfo generated from NodeTemplate
+		// If not a NodeTemplate, Pods from PodTemplates should alreadt be already present
+		// in the PodList attached to the Node.
+		if strings.HasPrefix(nodeInfo.Node().Name, "template-node-for-") {
+			var err error
+			newNodeInfo, err = getNodeInfoWithPodTemplates(nodeInfo, podTemplates, clusterSnapshot, ctx.PredicateChecker)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		} else {
+			newNodeInfo = nodeInfosForNodeGroups[id]
+		}
+
+		result[id] = newNodeInfo
+	}
+
+	return result, utilerrors.NewAggregate(errs)
+}
+
+// CleanUp cleans up processor's internal structuxres.
+func (p *nodeInfoWithPodTemplateProcessor) CleanUp() {
+	p.cancelFunc()
+}
+
+func newPodTemplateLister(kubeClient client.Interface, stopchannel <-chan struct{}) v1lister.PodTemplateLister {
+	podTemplateWatchOption := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.Everything().String()
+		options.LabelSelector = labels.SelectorFromSet(getDaemonsetPodTemplateLabelSet()).String()
+	}
+	listWatcher := cache.NewFilteredListWatchFromClient(kubeClient.CoreV1().RESTClient(), "podtemplates", apiv1.NamespaceAll, podTemplateWatchOption)
+	store, reflector := cache.NewNamespaceKeyedIndexerAndReflector(listWatcher, &apiv1.PodTemplate{}, time.Hour)
+	lister := v1lister.NewPodTemplateLister(store)
+	go reflector.Run(stopchannel)
+	return lister
+}
+
+func getNodeInfoWithPodTemplates(baseNodeInfo *schedulerframework.NodeInfo, podTemplates []*apiv1.PodTemplate, clusterSnapshot *simulator.BasicClusterSnapshot, predicateChecker simulator.PredicateChecker) (*schedulerframework.NodeInfo, error) {
+	node := baseNodeInfo.Node()
+	var pods []*apiv1.Pod
+
+	for _, podInfo := range baseNodeInfo.Pods {
+		pods = append(pods, podInfo.Pod)
+	}
+	if err := clusterSnapshot.AddNodeWithPods(node, pods); err != nil {
+		return nil, err
+	}
+	// clone baseNodeInfo to not modify the input object.
+	newNodeInfo := baseNodeInfo.Clone()
```

Comment:
DeepCopyTemplateNode from utils/scheduler? IIRC there were some issues with nodeInfo.Clone() being shallower copy than expected.

Reply:
good to know.
I will use the `DeepCopyTemplateNode` instead.